### PR TITLE
[theories]: p_max (maximum probability for atomic events)

### PR DESCRIPTION
**Pull request by chdoc**
This is the theory of `op p_max (d : 'a distr) = flub (mu1 d)` the maximal probability of atomic events in the distribution. 

This includes a proof that, despite the definition using `flub`, the maximum is always attained. This in turn required a few more lemmas on `flub` and factoring out the lemma:
```
lemma mu_pos_fin (d : 'a distr) eps : 
  0%r < eps => is_finite (fun x => eps <= mu1 d x).
```
from the proof of `uniform_finite`. 

`p_max` implicitly captures min-entropy. I chose to not include the latter in this PR, because the PR is already quire big.

There's quite a few names and places (for lemmas) that I'm not sure about. Please review carefully.

### CODE REVIEW

**Comment by fdupress** [comment]
The smt failures should be addressed by #355 if it doesn't get blackballed.

**Review by fdupress — APPROVED** [review]

File: theories/prelude/Logic.ec [hunk 1]
```diff
@@ -678,6 +678,11 @@ lemma semptyNP ['a] (E : 'a -> bool) :
   !sempty E <=> exists x, E x.
 proof. by rewrite /sempty -negb_exists. qed.
 
+(* -------------------------------------------------------------------- *)
+lemma pred_0Vmem (p : 'a -> bool) : p = pred0 \/ exists x, p x by smt().
```
Comment (fdupress) [inline theories/prelude/Logic.ec:682]: @strub can you please confirm that this is where this lemma belongs? We're not sure how to split material between the various prelude theories.

**Comment by fdupress** [comment]
We will be merging this one over #240 , but have marked it as co-authored given the joint origin of both in Ethan's Dilithium work.
Before merging, [this comment]9https://github.com/EasyCrypt/easycrypt/pull/346/commits/961fc3d1cc463925afc6617d460e6b48bade27a6#r1198942026) should be addressed (or actively dismissed) by @strub.

File: theories/prelude/Logic.ec (hunk 1 above)
Comment (strub) [reply theories/prelude/Logic.ec:682]: I would say `Core.ec`

Comment (fdupress) [reply theories/prelude/Logic.ec:682]: @chdoc, can you please move it (next to all the other lemmas about the `predX` operators) and amend/squash?

Comment (chdoc) [reply theories/prelude/Logic.ec]: done :heavy_check_mark:

**Comment by fdupress** [comment]
Sorry—this slipped between the virus-induced cracks. Let me take a quick look and merge.

**Pull request merged**